### ISSUE TITLE
v1.2.1: Low mesh peers due to p7

### ISSUE THREAD

**Issue by twoeths** [ChainSafe/lodestar#4818]
**Describe the bug**

Gossipsub score has been low since v1.2.1

<img width="1274" alt="Screen Shot 2022-11-27 at 12 54 15" src="https://user-images.githubusercontent.com/10568965/204121363-ebcb2586-3006-4674-a510-be07f3b0956d.png">

<img width="784" alt="Screen Shot 2022-11-27 at 12 55 03" src="https://user-images.githubusercontent.com/10568965/204121386-d52a33af-bb2f-46e3-8b70-eedb1f45acce.png">

due to P7

<img width="1570" alt="Screen Shot 2022-11-27 at 12 55 23" src="https://user-images.githubusercontent.com/10568965/204121392-b708ddd9-6c47-4be9-a53c-401c6837f046.png">

**Expected behavior**

Mesh peers should be the same to v1.1.0

**Comment by twoeths** [comment]
when I zoom in a range of `6h` using `1m` rate, it shows that if messages are delivered they're always within 6s. It means peers get penalties because they don't deliver messages they promise at all (or within 30s)?! Note that gossipsub delete tracked messages after 30s

<img width="1596" alt="Screen Shot 2022-11-29 at 08 38 23" src="https://user-images.githubusercontent.com/10568965/204417171-3b654d27-bb6b-4284-ad90-284f0275cf8f.png">

Update: this metric is the same to v1.1.1

this is not likely an I/O lag issue

**Comment by twoeths** [comment]
lodestar v1.2.1 actually does not subscribe to all subnets (even with a node of >= 64 validators)

<img width="1240" alt="Screen Shot 2022-11-29 at 17 44 32" src="https://user-images.githubusercontent.com/10568965/204508771-c5f29f4b-f265-4165-9484-a49e400ba047.png">

somehow a subnet topic is subscribed then unsubscribed all the times

<img width="1266" alt="Screen Shot 2022-11-29 at 17 45 35" src="https://user-images.githubusercontent.com/10568965/204508971-f6f89c62-9044-4d30-8790-5d84c1110c8f.png">

**Comment by twoeths** [comment]
Lodestar unsubscribe and then subscribe same subnet frequently

```
Nov-29 10:24:23.005[network]       ^[[36mverbose^[[39m: Unsubscribe to gossipsub topic topic=/eth2/4a26c58b/beacon_attestation_50/ssz_snappy
Nov-29 10:24:23.005[network]       ^[[36mverbose^[[39m: Unsubscribe to gossipsub topic topic=/eth2/4a26c58b/beacon_attestation_52/ssz_snappy
Nov-29 10:24:23.005[network]       ^[[36mverbose^[[39m: Subscribe to gossipsub topic topic=/eth2/4a26c58b/beacon_attestation_52/ssz_snappy
Nov-29 10:24:23.005[network]       ^[[36mverbose^[[39m: Subscribe to gossipsub topic topic=/eth2/4a26c58b/beacon_attestation_50/ssz_snappy
```

**Comment by twoeths** [comment]
the gossipsub 4.1.1 `unsubscribe()/leave()` is async and not finished so it ignores the `subscribe()` call if we do it immediately https://github.com/ChainSafe/js-libp2p-gossipsub/blob/v4.1.1/src/index.ts#L1680

for the context, in gossipsub 0.14.1 `leave()` method is sync, that's why issue did not happen in lodestar v1.1.1 https://github.com/ChainSafe/js-libp2p-gossipsub/blob/v0.14.1/ts/index.ts#L1602

from lodestar side, it should not unsubscribe() then subscribe() immediately

**Comment by twoeths** [comment]
Gossipsub broken promises on stable lg1k after we deploy v1.2 on Nov 09:

<img width="1286" alt="Screen Shot 2022-11-30 at 16 48 16" src="https://user-images.githubusercontent.com/10568965/204763377-7e31d30d-16b0-47a5-960e-211611526f7f.png">

**Comment by wemeetagain** [comment]
Take a look at the metrics on the latest libp2p branch. Seems p7 issue is resolved there

**Comment by twoeths** [comment]
a lot of broken promises are due to the collision of fastMsgId function, i.e. 2 different messages have the same fastMsgId. Right now tracer ignores duplicate messages, we should also track them as delivered, see https://github.com/ChainSafe/js-libp2p-gossipsub/pull/385

**Comment by wemeetagain** [comment]
Resolved with v1.3.0

**Comment by twoeths** [comment]
to be clear, this is fixed in curent `unstable` after libp2p is upgraded, so v1.4.0 not v1.3.0
